### DISCUSSION TITLE
New feature: Jekyll style post dating

### DISCUSSION THREAD

**Discussion by flyinggrizzly** [srid/emanote#440] [Ideas]
It would be useful to be able to create a page, say at `blog/2023-05-24-a-blog-post.md`, and have it infer its frontmatter date from the filename.

This [is the default and expected behavior in Jekyll for writing posts](http://jekyllrb.com/docs/posts/), and has two benefits (as I see it):

1. I can view posts on the filesystem, and know at a glance the publication sequence/how old a given post is without having to open it
    a. an added benefit is that all my posts are always sequentially sorted by date by the filesystem, and I don't need to open an editor and hop around
2. I don't have to explicitly date posts in frontmatter

I know Emanote isn't blog-centric in the way that Jekyll is, but this would improve ergonomics for me at least, and I imagine for anyone else familiar with Jekyll.

It also feels relatively low-risk as an ehancement, since the likelihood of someone happening to have a pile of pages named with dates in the filename and not desiring them to have those dates set seems *very* unlikely.

**NB**:

I believe Jekyll allows (or allowed) date overriding with an explicit frontmatter date. In those cases the frontmatter took precedence over the filename date.

**Comment by srid** [comment]
Have you explored writing a lua filter for it? https://emanote.srid.ca/guide/lua-filters

I don't recall off the top of my head if it also allows transforming the frontmatter metadata. If not, perhaps we can make it to be so.

**Reply by flyinggrizzly** [reply]
I haven't had a chance to try that, nor will I for a bit but... if the Lua filters can parse the filename, and have the ability to write frontmatter then that would probably solve this yea.

It would be really nice if we had the ability to set a filter for all files in a given path rather than having to specify it per-file if I'm understanding the current limitations correctly, but I also see that that's likely a temporary issue.

Thanks for the input!

**Comment by flyinggrizzly** [comment]
I've taken a swing at it, and it looks like we don't have the filename in the Lua filter process:

https://github.com/flyinggrizzly/two.flyinggrizzly.net/pull/1

<img width="1285" alt="image" src="https://github.com/srid/emanote/assets/10967676/7e15a8d6-0753-4326-9e73-982889e697bc">
☝️ there's a `print(doc)` there that shows no file/path data.

Am I looking at the wrong object in the filter (can't find anything higher than the `Pandoc` object itself in the docs), or would new work need to be done to expose it here?

Edit: I should point out also that `PANDOC_STATE.input_files` is an empty list too.

**Reply by srid** [reply]
> would new work need to be done to expose it here?

Probably yea, happy to accept a PR. Lua support is WIP after all: https://github.com/srid/emanote/issues/263

The code for this is here: 
- https://github.com/srid/emanote/blob/master/emanote/src/Emanote/Model/Note/Filter.hs
- Maybe you can pass the filepath here, https://github.com/srid/emanote/blob/5623bb3814e382dfea7ffd21d3da25c2d3179001/emanote/src/Emanote/Model/Note.hs#L334

**Reply by flyinggrizzly** [reply]
👍  I'll start looking at it. Haven't really dealt much with Haskell before so it'll be slow.
